### ISSUE TITLE
Lichess with Arduino Nano 33 IoT

### ISSUE THREAD

**Issue by vovagorodok** [vovagorodok/ArduinoBleChess#1]

**Comment by TimoKropp** [comment]
I tried running your AVR/ESP32 examples but I am not sure where the incomaptibiltiy to the SAMD21 lays. It compiles and runs but does not show during the scan process of the app (it shows however when searching for bluetooth devices). This is the same result which I was able to generate with the following code: `
#include <ArduinoBLE.h>

// cecp_protocol
#define SERVICE_UUID  "f5351050-b2c9-11ec-a0c0-b3bc53b08d33"
#define RX_UUID  "f53513ca-b2c9-11ec-a0c1-639b8957db99"
#define TX_UUID  "f535147e-b2c9-11ec-a0c2-8bbd706ec4e6"

BLEService OCBService(SERVICE_UUID); // create openchessboard(OCB)service

BLEStringCharacteristic OCBCharacteristic_RX(RX_UUID, BLEWrite, 64);
BLEStringCharacteristic OCBCharacteristic_TX(TX_UUID, BLERead | BLENotify, 64);

void setup() {
  Serial.begin(9600);
  while (!Serial);

  // begin initialization
  if (!BLE.begin()) {
    Serial.println("starting Bluetooth® Low Energy module failed!");
    while (1);
  }

  // set the local name peripheral advertises
  BLE.setLocalName("OPENCHESSBOARD");
  BLE.setDeviceName("OPENCHESSBOARD"); 
  // set the UUID for the service this peripheral advertises:
  BLE.setAdvertisedService(OCBService);

  // add the characteristics to the service
  OCBService.addCharacteristic(OCBCharacteristic_RX);
  OCBService.addCharacteristic(OCBCharacteristic_TX);

  // add the service
  BLE.addService(OCBService);

  // start advertising
  BLE.advertise();

  Serial.println("Bluetooth® device active, waiting for connections...");
}

void loop() {
  BLEDevice central = BLE.central();
  
  if (central) {
    Serial.print("Connected to central: ");
    // print the central's BT address:
    Serial.println(central.address());
  }
  delay(1000);

}`

**Comment by vovagorodok** [comment]
Example optimized for AVR, by removing STL. Hmm, strange. Code that you show looks OK. I had similar issue in the past with ESP32. Issue was because peripheral device has service, but doesn't advertise it. Line solves issue:
```
advertising->addServiceUUID(SERVICE_UUID);
```
As I see You have added similar line:
```
BLE.setAdvertisedService(OCBService);
```
Perhaps order of this line can change behavior. Try:
```
// start advertising
BLE.setAdvertisedService(OCBService);
BLE.advertise();
```
Please check in `nRF Connect` if your device correctly advertise uuid:
<img src="https://user-images.githubusercontent.com/8137447/170259368-07f417c0-b0b5-482d-ae1d-ed3ffc011d52.png" width="350"/><img src="https://user-images.githubusercontent.com/8137447/170259584-f2a0e251-6ca6-4ff8-914c-79c4be5c488c.png" width="350"/>

**Comment by TimoKropp** [comment]
My example shows in nRF connect identical as you have shown in the images.

**Comment by TimoKropp** [comment]
In the Scanner window (right image) it shows "**incomplete** List of 128-Bit Servive UUIDs:" and then shows the correct UUID
instead of **complete list**

**Comment by vovagorodok** [comment]
Just for double check. Have You enabled bluetooth and localization ? Because its required, and if was disabled scan doesn't find anything

**Comment by TimoKropp** [comment]
> 

Yes it is enabled.

**Comment by vovagorodok** [comment]
And lichess app has localization permissions?

**Comment by TimoKropp** [comment]
> And lichess app has localization permissions?

yes, that is what i checked first :)

**Comment by vovagorodok** [comment]
And you have visible UUID in `FLAGS & SERVICES` tab in nRF Connect?

**Comment by TimoKropp** [comment]
okay I tested it on a different device  with a fresh installation and now it shows during scan. Maybe there was an issue during the apk installation.

**Comment by vovagorodok** [comment]
Nice!

**Comment by TimoKropp** [comment]
okay it seems to allow a connection during scan (works with both code examples) but when i go back to the main menu in the app, it repetitively shows "disconnect from Bluetooth device "

**Comment by TimoKropp** [comment]
> For both code examples?

Your AVR/ESP32 example and my posted example.

**Comment by TimoKropp** [comment]
After your recent commits the error **ble initialization error** is shown. I think there went something wrong

**Comment by vovagorodok** [comment]
Ouu, I think that because of typo:
```
if (!begin(device));
{
```
insted:
```
if (!begin(device))
{
```
Fixed

**Comment by TimoKropp** [comment]
Init works now but still no debug output in serial monitor and the app shows repetitive disconnects.

**Comment by vovagorodok** [comment]
Added `Serial.println("setup");` and `Serial.println("loop");`. Please check if it reboots, an You  see `setup` log each time

**Comment by TimoKropp** [comment]
Setup is only called once, so no device restart. It repetitively posts loop but nothing else.

**Comment by vovagorodok** [comment]
Added more logs, please check

**Comment by TimoKropp** [comment]
`begin: BLE: 20000274
begin: ArduinoBleChessClass: 20000244
begin: Protocol: 2000026C
CecpProtocol::begin: Protocol: 2000026C
setup: BLE: 20000274
setup: ArduinoBleChessClass: 20000244
`
still no prints during scan

**Comment by vovagorodok** [comment]
But device advertises and You find it during scan? Look that nothing received via BLE. Added logs for connected/discon./discover events and checking i any data on characteristic is available

**Comment by TimoKropp** [comment]
> If You want we can try call via skype or another communicator in order to boost issue resolving

let me see if I find the time next week :)

**Comment by vovagorodok** [comment]
> still no changes, the new logs don't show up. The device advertise works since it shows up during scan but nothing more from that. Maybe the issue is on app side?
Hmm, hard to say. Works correctly with ESP32 and RaspberryPi (checked also with apk that was attached)

**Comment by TimoKropp** [comment]
If ESP32 works but not AVR, I would start by comparing the difference in both implementations. I think the device somehow stops listening after advertisment, or does not receive anything after advertisment (very unlikely).

**Comment by TimoKropp** [comment]
For example
`auto* advertising = server->getAdvertising();
    advertising->setScanResponse(true);
    advertising->setMinPreferred(0x06); // functions that help with iPhone connections issue
    advertising->setMaxPreferred(0x12);`
This seems quite different to the AVR example.

**Comment by TimoKropp** [comment]
Also the example in the ArduinoBLE library starts with an initial value for an characteristics:
https://github.com/arduino-libraries/ArduinoBLE/blob/master/examples/Peripheral/CallbackLED/CallbackLED.ino

Maybe this is missing this causes the app to immediately disconnect after the discovery?

But I am only guessing since I don't have that much time to debug currently.

**Comment by TimoKropp** [comment]
Also the example polls permanently in loop :(

**Comment by vovagorodok** [comment]
Perhaps ArduinoBLE implementation require pull in loop(). Added BLE.poll() to example

**Comment by TimoKropp** [comment]
Okay nice, polling works. It seems that the library does rely on polling. Not the best but I think it is okay for the chess board application since it not that fast.

**Comment by vovagorodok** [comment]
Ok, good to know. Updated library. In future `CecpProtocol` class will be refactored to state machine. What will be nice to have or what to remove in such library from your PoV? Interface will be changed until v1.0.0, because I'll test it on real device. And will be added to PlatformIO at the end

**Comment by TimoKropp** [comment]
There are already some issues with the current implementation which a state machine could solve:
1. when starting a game as black, the last move is only in the transmitted fen and it would be more useful to simply transmit the last move as move string. Only when canceling the game, the last move is indicated as "moved from phone".
2. When a game is already running in the app and the device is reconnected, no game data is received. There again at least the last move needs to be requested/transmitted
3. When making a move on the phone, the move should be broadcasted to the device so you can update your board state on the smart chess board.

There are some issues with the app:
1. When enabling and disabling bluetooth in the app, no new scan is run and there is no indication of a connected device. Only a restart of the app allows a full new scan.
On my other Device ( Huawei Tablet) the rescan works as the switch does disconnect the bluetooth connection.
I guess that is an issue from the installation and app permission but it is not clear how to solve this in the nokia case. The permissions are correctly indicated in the app manager.
2. When no device found during scan, the switch is not set back to off.
3. When using the tablet, the device is stuck at "stop move: ", not clear why.

**Comment by TimoKropp** [comment]
Further if you try to login to a user account, the app crashes.

**Comment by vovagorodok** [comment]
Nice to see such precise testing

Let's see:

DEVICE SIDE:


1. Perhaps I don't understand usecase. When I'm trying play with computer from screach as black:
```
new game: rnbqkbnr/pppppppp/8/8/8/8/PPPPPPPP/RNBQKBNR w
moved from phone: d2d4
please move: 
```
Initial board state is received and wite move (last move) is received after that. If you play not from screach board already have recived moves before and we just recive current board state. If we will add additional move it will be redundant information and I think it's not ok to send same move twice. Additionaly it's not compatible with the XBoard protocol. Please describe more usecase if I didn't understand

2. Hmm, how to reproduce this case? During disconnection from board there can be more than one move performed and I think that send only one last move does't solve that case. How do you think we can solve it? Send new game and fen when we connected again and chess board will compare what was lost during disconnection? 

3. Already solved by protocol. Look: https://github.com/vovagorodok/lichobile/blob/master/src/externalDevice/README.md case: `When user performs a2a3 move from screen:`. On protocol side we will see:
```
-> go     // ask device to move
-> force  // ask device to stop move, becasuse moved on screen
-> a2a3   // send move performed on screen
-> a7a6   // send oponent move
-> go     // ask device to move
```
In logs you will see:
```
please move: 
stop move: 
moved from phone: b7b6
moved from phone: g1f3
please move: 
```


PHONE SIDE:


1. Works as expected on all my devices. Perhaps It's because apk builded not correctly.

2. Known issue :p Fixed: https://github.com/lichess-org/lichobile/pull/2101/commits/a921dff29b03751e44006b096b429cc22a7d95c4

3. Hard to say what happens without full usecase description (parhaps it because apk not ok)


> Further if you try to login to a user account, the app crashes.

I'm C++/Python dev and it's my first experience with mobile/java. I guess that because apk is not correctly build, and not correctlly support online play. Perhaps will be nice to ask guys from lichobile build full apk for us

**Comment by TimoKropp** [comment]
test case 1:
game was started and its my turn with white (1 move played by black). After scan and connect, the following is shown:
`moved from phone: xboard
please move:
moved from phone: accepted setboard
stop move:`

**Comment by vovagorodok** [comment]
Fixed: https://github.com/vovagorodok/ArduinoBleChess/commit/34156a7d854ca13a2ad5a9faa6b1ab17a8e67f1b

Works in ESP because regexp of uci move, planed to add regexp for ArduinoBLE case

**Comment by TimoKropp** [comment]
It seems that the last move is only received on device side after the game was canceled/resigned in the app (reproducible)

**Comment by TimoKropp** [comment]
when closing the app and opening the game again:
`new game: rnbqkbnr/pp1p2pp/2p1p3/5p2/3PPP2/8/PPP3PP/RNBQKBNR w
please move: `
the last move is not received.

**Comment by vovagorodok** [comment]
> when closing the app and opening the game again: `new game: rnbqkbnr/pp1p2pp/2p1p3/5p2/3PPP2/8/PPP3PP/RNBQKBNR w please move: ` the last move is not received.

Yes, because last move was recived before closing app. Correct?

**Comment by TimoKropp** [comment]
aaah now I get the difference. There is an issue with playing offline/computer compared to with play with the computer over the main menu (I guess this is another endpoint).
The offline version seems to work but the computer/online game does not.

**Comment by vovagorodok** [comment]
> Yes, because last move was recived before closing app and device was already triggered to move . Correct?

But I don't get what the issue is here

**Comment by TimoKropp** [comment]
> > Yes, because last move was recived before closing app and device was already triggered to move . Correct?
> 
> But I don't get what the issue is here

Yes it is an issue with the online play against the computer. It works well if you only use the offline board.

**Comment by TimoKropp** [comment]
Do you try to support online play if this is permitted by the lichobile devs or do you only want to use offline play? I think the white pawn app does exactly that to allow online play. I would strongly prefer to play online vs. offline games :)

**Comment by vovagorodok** [comment]
Ok, lets summarize issues that we need to solve in order to simply track it:
1. When a game is already running in the app and the device is reconnected, no game data is received.  How to reproduce it?
2. When using the tablet, the device is stuck at "stop move: ", not clear why. Reproduced and under investigation
3. Online play. Was added, but I can't build apk with online play and I don't know why. Lets as lichobile devs to prepare full apk

Something missed?

**Comment by TimoKropp** [comment]
> Ok, lets summarize issues that we need to solve in order to simply track it:
> 
>     1. When a game is already running in the app and the device is reconnected, no game data is received.  How to reproduce it?
I receive the FEN but not the last move played. How to test it:
1. start a game and play some moves
2. restart/reprogram peripheral device
3. reconnect bluetooth
4. received game data:
`moved from phone: xboard
please move:
moved from phone: accepted setboard
stop move: 
new game: rnbqkb1r/pp1ppp1p/6p1/2p5/3Pn3/1P3N2/P1P2PPP/RNBQKB1R w
please move: `

**Comment by vovagorodok** [comment]
> > Ok, lets summarize issues that we need to solve in order to simply track it:
> > ```
> > 1. When a game is already running in the app and the device is reconnected, no game data is received.  How to reproduce it?
> > ```
> 
> I receive the FEN but not the last move played. How to test it:
> 
>     1. start a game and play some moves
> 
>     2. restart/reprogram peripheral device
> 
>     3. reconnect bluetooth
> 
>     4. received game data:
>        `moved from phone: xboard please move: moved from phone: accepted setboard stop move:  new game: rnbqkb1r/pp1ppp1p/6p1/2p5/3Pn3/1P3N2/P1P2PPP/RNBQKB1R w please move: `

`moved from phone: xboard` already fixed on latest ArduinoBleChess. Are You updated?

**Comment by TimoKropp** [comment]
now updated:
`moved from phone: accepted setboard
please move: 
stop move: 
new game: rnbqkb1r/pp1ppp1p/6p1/2p5/3Pn3/1P3N2/P1P2PPP/RNBQKB1R w
please move: 
`
still not the last move shown.

**Comment by vovagorodok** [comment]
Fixed. Please update.

**Comment by vovagorodok** [comment]
what :) and you see: `moved from phone: accepted ` ? on latest version?

**Comment by TimoKropp** [comment]
I imported the projectv again and this is the only result:

`new game: rnbqkbnr/pp2pppp/2p5/3p4/4P3/5N2/PPPP1PPP/RNBQKB1R w
please move: `

that is all I see after a peripheral device reprogramming and then opening the ongoing game in the app.

**Comment by TimoKropp** [comment]
I guess the last move was already send and received previously to/by the peripheral device. But I would suggest that it will be resend if you click on the game again in the app. So a device restart is able to catch the last move even if it was send from the app previously.

**Comment by TimoKropp** [comment]
I agree that the protocol needs to be generic/universal. The issue with the peripheral device is that most of the time you only store the move history in ram. This info is lost at a restart. If you want to keep the moves stored permanently you need to write to the flash which is not the most elegant way :(. Maybe you have an idea to request the last move separately outside of the standard protocol? This would also be useful for analysis mode of the chess board to be able to play through lines requested by the board.

**Comment by vovagorodok** [comment]
Understand. For me is ok to save one move on flash/eeprom. I'll save even more (options, last board state if we play with AI inside board). Checked protocol and there no such options, because protocol is not hardware related and I thing shouldn't be. Each additional command/exception will complicate it and I think we should avoid that. Using Arduino EEPROM library for save one move will be most elegant way as I see

**Comment by TimoKropp** [comment]
I am not sure about that, the EEPROM has 100,000 write/erase cycles which could be an issue for the device lifetime. Most users would never play/abort that much games but you never know :). But I think it is okay to lose the last move functionality currently or live the limited live time.

**Comment by TimoKropp** [comment]
I am in contact with Erik from chess.com and he is interested in adding compatibility for the cecp protocol in the chess.com app. Would you mind if I take you on the email exchange? Just link me your email and I will take you in for the discussion.

**Comment by vovagorodok** [comment]
Nice step! yes of course

**Comment by vovagorodok** [comment]
> 2. When using the tablet, the device is stuck at "stop move: ", not clear why. Reproduced and under investigation

Nice finding. Issue reported to https://github.com/capacitor-community/bluetooth-le/issues/341
Now under discussion, which is better way to solve it

**Comment by TimoKropp** [comment]
I also put you in the discussion with Khadim, to work on an integration to the white pawn app ;)

**Comment by TimoKropp** [comment]
Still getting permanently disconnects on my Nokia 7.2 after successful testing yesterday. It seems very unstable for this device. Not clear why.

**Comment by vovagorodok** [comment]
> Still getting permanently disconnects on my Nokia 7.2 after successful testing yesterday. It seems very unstable for this device. Not clear why. Now I also get disconnects with Huawei Tablet. solved: I broke something :)

Ok, no issue here?

A I understand currently we have two issues to solve:
1.  Stuck at "stop move: ". Under discussion wit BLE library authors
2. Online play. Waiting response from lichobile guys

**Comment by TimoKropp** [comment]
Yes, those are the open points. Still no answer from the lichobile devs unfortunately :(. In the meentime, I posted a video of the Bluetooth integration on my discord. 
https://discord.gg/MFmSyHWzbK
It works really well with your Bluetooth driver :)

**Comment by vovagorodok** [comment]
Love it! You can show it with information about responsiveness and blitz possibility in lichobile topic
